### PR TITLE
Update K8s docker image build and the source artifact registry

### PR DESCRIPTION
**Pull request by yika-luo**
This PR depends on https://github.com/skypilot-org/skypilot-catalog/pull/98

Impact of this change:
1. Make `setup.py` the single source of truth of sky dependencies 
2. Host images in skypilot's own GCP projects. 
3. Performance gain: ~10s speed up in k8s sky cluster creation

|  VM Type 💻   | Old Provision 🕐                 | New Provision 🕐     | % Speedup ✅ |
| --------------- | ---------------------------- | ----------- | ----------- |
| K8 CPU      |   58s    | 34s   | 40% (2x) |
| K8 GPU | 42s | 34s | 20% (1.5x)|

Tested (run the relevant ones):
- [x] Made new artifact repos public
- [x] Manual test for both CPU and GPU: `./skypilot-k8s-image.sh -p -l -r ${REGION}` --> change local catalog --> `sky local up` --> `sky launch --cloud=kubernetes`
- [x] Minimal smoke test: `pytest tests/test_smoke.py::test_minimal --kubernetes`

### CODE REVIEW

File: Dockerfile_k8s [hunk 1]
```diff
@@ -33,13 +32,24 @@ ENV HOME /home/sky
...
+
+# Install python system packages
+RUN curl -o Miniconda3-Linux-x86_64.sh https://repo.anaconda.com/miniconda/Miniconda3-py310_23.11.0-2-Linux-x86_64.sh && \
+    bash Miniconda3-Linux-x86_64.sh -b && \
+    eval "$(~/miniconda3/bin/conda shell.bash hook)" && \
+    conda init && conda config --set auto_activate_base true && conda activate base && \
+    grep "# >>> conda initialize >>>" ~/.bashrc || { conda init && source ~/.bashrc; } && \
+    rm Miniconda3-Linux-x86_64.sh && \
+    conda create -y -n skypilot-runtime python=3.10 && \
+    conda activate skypilot-runtime && \
```
Comment (Michaelvll) [inline Dockerfile_k8s:45]: We don't need to install conda or creating a new env for skypilot-runtime with conda, as we are already using a conda base image. The lines below show create the venv for `skypilot-runtime` we need

File: Dockerfile_k8s [hunk 2]
```diff
@@ -33,13 +32,24 @@ ENV HOME /home/sky
...
+    export PIP_DISABLE_PIP_VERSION_CHECK=1 && \
+    python3 -m venv ~/skypilot-runtime && \
+    PYTHON_EXEC=$(echo ~/skypilot-runtime)/bin/python && \
+    $PYTHON_EXEC -m pip install 'skypilot-nightly[remote,docker,kubernetes]'
```
Comment (Michaelvll) [inline Dockerfile_k8s:49]: Do we need the docker dependency here?
```suggestion
    $PYTHON_EXEC -m pip install 'skypilot-nightly[remote,kubernetes]'
```

File: Dockerfile_k8s [hunk 3]
```diff
@@ -33,13 +32,24 @@ ENV HOME /home/sky
...
+
+# Install kubectl
+RUN curl -LO "https://dl.k8s.io/release/v1.28.11/bin/linux/amd64/kubectl" && \
     sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl
```
Comment (Michaelvll) [inline Dockerfile_k8s:53]: Let's keep the curl in the same RUN above so we can keep it in the same layer?

File: Dockerfile_k8s_gpu [hunk 4]
```diff
@@ -36,16 +36,21 @@ SHELL ["/bin/bash", "-c"]
...
+    conda init && conda config --set auto_activate_base true && conda activate base && \
     grep "# >>> conda initialize >>>" ~/.bashrc || { conda init && source ~/.bashrc; } && \
     rm Miniconda3-Linux-x86_64.sh && \
-    pip install wheel Click colorama cryptography jinja2 jsonschema networkx  \
-    oauth2client pandas pendulum PrettyTable rich tabulate filelock packaging  \
-    'protobuf<4.0.0' pulp pycryptodome==3.12.0 docker kubernetes==28.1.0  \
-    grpcio==1.51.3 python-dotenv==1.0.1 ray[default]==2.9.3 && \
-    curl -LO "https://dl.k8s.io/release/v1.28.11/bin/linux/amd64/kubectl" && \
+    conda create -y -n skypilot-runtime python=3.10 && \
+    conda activate skypilot-runtime && \
```
Comment (Michaelvll) [inline Dockerfile_k8s_gpu:46]:
```suggestion
```
We should avoid creating a conda env here.

File: Dockerfile_k8s [hunk 5]
```diff
@@ -9,8 +9,7 @@ ARG DEBIAN_FRONTEND=noninteractive
 RUN apt update -y && \
     apt install git gcc rsync sudo patch openssh-server pciutils nano fuse socat netcat curl -y && \
     rm -rf /var/lib/apt/lists/* && \
-    apt remove -y python3 && \
-    conda init
+    apt remove -y python3
```
Comment (Michaelvll) [inline Dockerfile_k8s:12]: Keep the original `conda init`?

File: sky/clouds/kubernetes.py [hunk 6]
```diff
@@ -68,8 +68,11 @@ class Kubernetes(clouds.Cloud):
                                                              'Kubernetes.',
     }
 
-    IMAGE_CPU = 'skypilot:cpu-ubuntu-2004'
-    IMAGE_GPU = 'skypilot:gpu-ubuntu-2004'
+    IMAGE_CPU = 'skypilot:custom-cpu-ubuntu-2004'
+    IMAGE_GPU = 'skypilot:custom-gpu-ubuntu-2004'
```
Comment (Michaelvll) [inline sky/clouds/kubernetes.py:72]: Do we need to add the automatic refetching for the catalog as well?

File: tests/kubernetes/build_image.sh [hunk 7]
```diff
@@ -9,30 +9,33 @@
 # -p: Push the image to the registry
```
Comment (Michaelvll) [inline tests/kubernetes/build_image.sh:9]: Should we move this file to the `sky/clouds/service_catalog/images`

File: Dockerfile_k8s_gpu [hunk 8]
```diff
@@ -41,15 +41,14 @@ RUN curl https://repo.anaconda.com/miniconda/Miniconda3-py310_23.11.0-2-Linux-x8
...
+    $PYTHON_EXEC -m pip uninstall skypilot-nightly -y && \
     curl -LO "https://dl.k8s.io/release/v1.28.11/bin/linux/amd64/kubectl" && \
-    sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl
-
-# Add /home/sky/.local/bin/ to PATH
-RUN echo 'export PATH="$PATH:$HOME/.local/bin"' >> ~/.bashrc
+    sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl && \
+    echo 'export PATH="$PATH:$HOME/.local/bin"' >> ~/.bashrc
 
 # Copy SkyPilot code base. This is required for the ssh jump pod to find the
```
Comment (Michaelvll) [inline Dockerfile_k8s_gpu:53]: cc'ing @romilbhardwaj , should we remove this now?

File: Dockerfile_k8s [hunk 9]
```diff
@@ -33,17 +33,15 @@ ENV HOME /home/sky
...
     curl -LO "https://dl.k8s.io/release/v1.28.11/bin/linux/amd64/kubectl" && \
-    sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl
-
-# Add /home/sky/.local/bin/ to PATH
-RUN echo 'export PATH="$PATH:$HOME/.local/bin"' >> ~/.bashrc
+    sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl && \
+    echo 'export PATH="$PATH:$HOME/.local/bin"' >> ~/.bashrc
 
 # Copy SkyPilot code base. This is required for the ssh jump pod to find the
 # lifecycle management scripts
```
Comment (Michaelvll) [inline Dockerfile_k8s:47]: @romilbhardwaj same here.

**Review by Michaelvll — COMMENTED** [review]
Thanks for adding this @yika-luo! Could we test the speed up with `--system-site-packages` removed from our setup script in #4168? I suspect the speed up would be more significant.

Also, it is a bit surprising that the provisioning for CPU is longer than GPU. Is it because the CPU image is larger than GPU image? If so, maybe we should move to `ubuntu` base image instead.

**Comment by yika-luo** [comment]
> Thanks for adding this @yika-luo! Could we test the speed up with `--system-site-packages` removed from our setup script in #4168? I suspect the speed up would be more significant.
> 
> Also, it is a bit surprising that the provisioning for CPU is longer than GPU. Is it because the CPU image is larger than GPU image? If so, maybe we should move to `ubuntu` base image instead.

Oh I ran CPU on my macbook.. I ran again on the same GPU instance and refreshed the result. They look comparable now.
I also tested removing `--system-site-packages` and it only gained 1s, but at least no regression :)

File: sky/clouds/kubernetes.py [hunk 10]
```diff
@@ -319,10 +322,14 @@ def make_deploy_resources_variables(
...
             dryrun: bool = False) -> Dict[str, Optional[str]]:
         del cluster_name, zones, dryrun  # Unused.
+        meta_region = self.DEFAULT_IMAGE_META_REGION
         if region is None:
             context = kubernetes_utils.get_current_kube_config_context_name()
         else:
             context = region.name
+            img_meta_region = region.name.split('-')[0]
+            if img_meta_region in self.IMAGE_META_REGIONS:
+                meta_region = img_meta_region
```
Comment (Michaelvll) [inline sky/clouds/kubernetes.py:332]: Actually, the region name is just the context name of the kubernetes cluster, which won't contain the location information, so probably we have to always use us.

File: Dockerfile_k8s [hunk 11]
```diff
@@ -33,21 +33,15 @@ ENV HOME /home/sky
...
-RUN conda init && \
-    pip install wheel Click colorama cryptography jinja2 jsonschema networkx  \
-    oauth2client pandas pendulum PrettyTable rich tabulate filelock packaging  \
-    'protobuf<4.0.0' pulp pycryptodome==3.12.0 docker kubernetes==28.1.0  \
-    grpcio==1.51.3 python-dotenv==1.0.1 ray[default]==2.9.3 && \
+# Install skypilot dependencies
+RUN conda init && export PIP_DISABLE_PIP_VERSION_CHECK=1 && \
+    python3 -m venv ~/skypilot-runtime && \
+    PYTHON_EXEC=$(echo ~/skypilot-runtime)/bin/python && \
+    $PYTHON_EXEC -m pip install 'skypilot-nightly[remote,kubernetes]' 'ray[default]==2.9.3' 'pycryptodome>=3.12.0' && \
```
Comment (Michaelvll) [inline Dockerfile_k8s:40]: Just curious, any reason we change to `>=` here?

**Review by Michaelvll — APPROVED** [review]
Thanks for the pr @yika-luo! It looks mostly good to me. Just a quick question below

**Pull request merged**